common/lib/dependabot/file_parsers/base/dependency_set.rb

```diff
@@ -50,41 +60,101 @@ def case_sensitive?
-          return dependencies.find { |d| d.name == name } if case_sensitive?
+        def key_for_name(name)
+          case_sensitive? ? name : name.downcase
+        end
 
-          dependencies.find { |d| d.name&.downcase == name&.downcase }
+        def key_for(dep)
+          key_for_name(dep.name)
         end
 
-        def combined_dependency(old_dep, new_dep)
-          package_manager = old_dep.package_manager
-          v_cls = Utils.version_class_for_package_manager(package_manager)
-
-          # If we already have a requirement use the existing version
-          # (if present). Otherwise, use whatever the lowest version is
-          new_version =
-            if old_dep.requirements.any? then old_dep.version || new_dep.version
-            elsif !v_cls.correct?(new_dep.version) then old_dep.version
-            elsif !v_cls.correct?(old_dep.version) then new_dep.version
-            elsif v_cls.new(new_dep.version) > v_cls.new(old_dep.version)
-              old_dep.version
+        class DependencySlot
+          attr_reader :combined
+
+          def initialize
+            @all_versions = []
+            @combined = nil
+          end
+
+          def all_versions(sort:)
+            sort ? sorted : @all_versions.dup
+          end
+
+          def <<(dep)
+            return self if @all_versions.include?(dep)
+
+            @combined = if @combined
+                          combine(@combined, dep)
+                        else
+                          Dependency.new(
+                            name: dep.name,
+                            version: dep.version,
+                            requirements: dep.requirements,
+                            package_manager: dep.package_manager,
+                            subdependency_metadata: dep.subdependency_metadata
+                          )
+                        end
+
+            index_of_same_version =
+              @all_versions.find_index { |other| other.version == dep.version }
+
+            if index_of_same_version.nil?
+              @all_versions << dep
             else
-              new_dep.version
+              same_version = @all_versions[index_of_same_version]
+              @all_versions[index_of_same_version] = combine(same_version, dep)
+            end
+
+            self
+          end
+
+          private
+
+          def combine(old_dep, new_dep)
+            version = if old_dep.requirements.any?
+                        old_dep.version || new_dep.version
+                      elsif !version_class.correct?(new_dep.version)
+                        old_dep.version
+                      elsif !version_class.correct?(old_dep.version)
+                        new_dep.version
+                      elsif version_class.new(new_dep.version) > version_class.new(old_dep.version)
+                        old_dep.version
+                      else
+                        new_dep.version
+                      end
```

Comment:
Could you add a short comment about what this combine "algorithm" does and (more importantly) why? It's not immediately clear to me why this works the way it does.

Reply:
It's not immediately clear to me either 😂 but yes, absolutely.

I carried this over verbatim from the [existing code](https://github.com/dependabot/dependabot-core/blob/91112975d6748eb0d211dccd23ca3f97934b6d00/common/lib/dependabot/file_parsers/base/dependency_set.rb#L63-L73) in order not to break anything. One thing I don't like about it is the fact that it will have different outcomes depending on the order in which versions of dependency are added—for example, if the first added has `#requirements`, its version will always win out even if it's the highest version, otherwise the lowest version will always win out—but not new behavior (and things broke in various ecosystems if I tried to modify that).

Reply:
Updated in ab9036b3cff2294b0f5306f4e4f563aa77317c87

Reply:
> if the first added has `#requirements`, its version will always win out even if it's the highest version, otherwise the lowest version will always win out—but this isn't new behavior (and things broke in various ecosystems if I tried to modify that).

I think the intention there is that the version of a direct dependency should be preferred over transitive versions.

Reply:
Ah, thanks for that context. This should be using `Dependency#top_level?` then so it's clear that that's the intent. Otherwise it's not really clear (to me at least) that direct vs. transitive is what the code is concerned with here.